warehouse/query-core/src/main/java/datawave/query/common/grouping/GroupingUtil.java

```diff
@@ -90,18 +90,18 @@ public GroupingInfo getGroupingInfo(Map.Entry<Key,Document> entry, Set<String> g
             int longest = longestValueList(fieldToFieldWithContextMap);
             for (int i = 0; i < longest; i++) {
                 Collection<GroupingUtil.GroupingTypeAttribute<?>> fieldCollection = new HashSet<>();
-                String currentGroupingContext = "";
+                String currentGroupingContext = convertGroupingContext(i);
                 for (String fieldListItem : expandedGroupFieldsList) {
                     log.trace("fieldListItem: {}", fieldListItem);
                     Collection<String> gtNames = fieldToFieldWithContextMap.get(fieldListItem);
                     if (gtNames == null || gtNames.isEmpty()) {
                         log.trace("gtNames: {}", gtNames);
                         log.trace("fieldToFieldWithContextMap: {} did not contain: {}", fieldToFieldWithContextMap, fieldListItem);
                     } else {
-                        String gtName = gtNames.iterator().next();
-                        int idx = gtName.indexOf('.');
-                        if (idx != -1) {
-                            currentGroupingContext = gtName.substring(idx + 1);
+                        String nameWithGrouping = fieldListItem + "." + currentGroupingContext;
+                        final String gtName = gtNames.stream().filter(name -> nameWithGrouping.equals(name)).findAny().orElse(null);
```

Comment:
The fieldToFieldWithContextMap is being created as a TreeMultimap which returns a SortedSet (in fact a NavigatableSet) on the get call.  So gtNames is actually a NavigatableSet which means that getNames.get(name) is all you need here!  Doing a stream can be extremely costly when the number of groups is very large (which is definitely the case sometimes).
If you want to make this more obvious, change the getFieldToFieldWithGroupingContextMap routine to return a TreeMultimap instead, change fieldToFieldWithContextMap to be a TreeMultimap as well and change gtNames to be a NavigatableSet.

Reply:
Please look a the latest commit.